drivers/sensor/ina23x/ina237.c

```diff
@@ -20,51 +20,69 @@ LOG_MODULE_REGISTER(INA237, CONFIG_SENSOR_LOG_LEVEL);
 #define INA237_CAL_SCALING 8192ULL
 
 /** @brief The LSB value for the bus voltage register, in microvolts/LSB. */
-#define INA237_BUS_VOLTAGE_UV_LSB 3125
+#define INA237_BUS_VOLTAGE_TO_uV(x) ((x) * 3125U)
 
-/** @brief Power scaling (scaled by 10) */
-#define INA237_POWER_SCALING 2
+/** @brief Power scaling (need factor of 0.2) */
+#define INA237_POWER_TO_uW(x) ((x) / 5ULL)
+
+/**
+ * @brief Scale die temperture from 0.125 degC/bit to micro-degrees C
+ *  Note that the bottom 4 bits are reserved and are always zero.
+ */
+#define INA237_DIETEMP_TO_uDegC(x) (((x) >> 4) * 125000)
+
+static void micro_s32_to_sensor_value(struct sensor_value *val, int32_t value_microX)
+{
+	val->val1 = value_microX / 1000000L;
+	val->val2 = abs(value_microX) % 1000000L;
+	if (value_microX < 0) {
+		val->val2 = -val->val2;
+	}
```

Comment:
This can be simplified and made about twice as fast if written as:
```c
    val->val1 = value_microX / 1000000L;
    val->val2 = value_microX % 1000000L;
```

The compiler can do one signed division this way, instead of an unsigned division and a signed division.

This will always generate the correct value for C99 or later.  Zephyr is C11 minimum.  In C89, it may or may not generate the correct output, depending on the CPU, but what you have now will also be wrong for the same reason.

In C99, division that produces a negative result rounds toward zero, e.g. -1200000 / 100000 = -1.  In C89, it is implementation defined if it rounds toward zero or rounds down, e.g. -1200000 / 100000 = -2.

So in C89, **val1** could have the wrong result (-2, rather than -1).  But thankfully it's been 24 years since that was addressed and we don't have to worry about it anymore.

In all C standards, `%` is defined in terms of division:  `(a / b) * b + (a % b) = a`

With `(a / b)` rounding toward zero, the result of `(a % b)` will need to be negative, e.g. -200000 in the above example, which is what we want.  If division rounded down, then it would need to be positive, 800000.

So your code does produce the correct `val2` even on the (unusual!) rounding-down C89 compiler, but it only makes a difference when `val1` will be incorrect, so the total result will still be wrong.

Reply:
Thanks the for explanation, will fix this.

Reply:
All fixed in the latest commit.  Thanks again for the detailed explanation since I have been forever thinking of % in C as modulus, so I learned something valuable from you.